### PR TITLE
feat: create informers for kube contexts

### PR DESCRIPTION
**Pull request by feloy**
Signed-off-by: Philippe Martin <phmartin@redhat.com>

### What does this PR do?

- adds a new boolean Preference entry `Get information about clusters declared in Kubeconfig`, to enable / disable the running of informers watching pods/deployments/replicasets for different kube contexts
- if enabled, the kubernetes client singleton manages a state of informers for the different contexts (stops informers of delete contexts, starts new informers for new contexts)
- if enabled, displays in the list of kube contexts if the cluster is reachable, and the number of current pods/deployemnts/replicasets for each kube context

### What issues does this PR fix or reference?

#4562

### How to test this PR?

- Enable the feature in the Preferences
- Switch contexts and check information is updated
- add/remove resources (pods, deployments) from context and check information is updated
-

### CODE REVIEW

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+      this.contextsInternalState.clear();
+      this.contextsState.clear();
+      this.dispatchContextsState();
+    }
+    console.log('==> # informers', this.contextsInternalState.size, this.contextsState.size);
+  }
+
+  private createKubeContextInformers(context: KubeContext): ContextInternalState | undefined {
+    const kc = new KubeConfig();
+    const cluster = this.kubeConfig.clusters.find(c => c.name === context.cluster);
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts:81]:
```suggestion
    const cluster = this.kubeConfig.clusters.find(c => c.name === context.cluster);
    if (!cluster) {
      retrun;
    }
```

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    const user = this.kubeConfig.users.find(u => u.name === context.user);
+    if (!cluster || !user) {
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]:
```suggestion
    if (!user) {
```

Comment (dgolovin) [reply packages/main/src/plugin/kubernetes-context-state.ts]: Why calculate user if there is no cluster?

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+    informer.on('delete', (_obj: V1Pod) => {
+      const previous = this.contextsState.get(context.name);
+      if (previous) {
+        previous.podsCount--;
+      }
+      this.dispatchContextsState();
+    });
+    informer.on('error', (err: unknown) => {
+      console.error(`==> ${context.name}: error`, err);
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]:
```suggestion
      console.error(`Pod informer error for context '${context.name}':`, String(err));
```

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        previous.reachable = err === undefined;
+      }
+      this.dispatchContextsState();
+      // Restart informer after 5sec
+      setTimeout(() => {
+        this.restartInformer(informer, context);
+      }, 5000);
+    });
+    informer.on('connect', (err: unknown) => {
+      console.error(`==> ${context.name}: connect`, err);
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]:
```suggestion
      console.error(`Pod informer connection error for context '${context.name}':`, String(err));
```

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+    informer.on('add', (_obj: V1Pod) => {
+      const previous = this.contextsState.get(context.name);
+      if (previous) {
+        previous.podsCount++;
+      }
+      this.dispatchContextsState();
+    });
+
+    informer.on('delete', (_obj: V1Pod) => {
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]:
```suggestion
    informer.on('delete', () => {
```
My understanding the 'underscore' would be good if you have first parameter that is not used and then some more that is used. If it is the only parameter, you can just skip it.

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+  private createReplicasetInformer(
+    kc: KubeConfig,
+    ns: string,
+    context: KubeContext,
+  ): Informer<V1ReplicaSet> & ObjectCache<V1ReplicaSet> {
+    const k8sApi = kc.makeApiClient(AppsV1Api);
+    const listFn = () => k8sApi.listNamespacedReplicaSet(ns);
+    const informer = makeInformer(kc, `/apis/apps/v1/namespaces/${ns}/replicasets`, listFn);
+
+    informer.on('add', (_obj: V1ReplicaSet) => {
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]:
```suggestion
    informer.on('add', () => {
```

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+      const previous = this.contextsState.get(context.name);
+      if (previous) {
+        previous.replicasetsCount++;
+      }
+      this.dispatchContextsState();
+    });
+
+    informer.on('delete', (_obj: V1ReplicaSet) => {
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]:
```suggestion
    informer.on('delete', () => {
```

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+    });
+
+    informer.on('delete', (_obj: V1Deployment) => {
+      const previous = this.contextsState.get(context.name);
+      if (previous) {
+        previous.deploymentsCount--;
+      }
+      this.dispatchContextsState();
+    });
+    informer.on('error', (_err: unknown) => {
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]:
```suggestion
    informer.on('error', () => {
```

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+    informer.on('add', (_obj: V1Deployment) => {
+      const previous = this.contextsState.get(context.name);
+      if (previous) {
+        previous.deploymentsCount++;
+      }
+      this.dispatchContextsState();
+    });
+
+    informer.on('delete', (_obj: V1Deployment) => {
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]:
```suggestion
    informer.on('delete', () => {
```

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+  private createDeploymentInformer(
+    kc: KubeConfig,
+    ns: string,
+    context: KubeContext,
+  ): Informer<V1Deployment> & ObjectCache<V1Deployment> {
+    const k8sApi = kc.makeApiClient(AppsV1Api);
+    const listFn = () => k8sApi.listNamespacedDeployment(ns);
+    const informer = makeInformer(kc, `/apis/apps/v1/namespaces/${ns}/deployments`, listFn);
+
+    informer.on('add', (_obj: V1Deployment) => {
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]:
```suggestion
    informer.on('add', () => {
```

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+    });
+
+    informer.on('delete', (_obj: V1ReplicaSet) => {
+      const previous = this.contextsState.get(context.name);
+      if (previous) {
+        previous.replicasetsCount--;
+      }
+      this.dispatchContextsState();
+    });
+    informer.on('error', (_err: unknown) => {
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]:
```suggestion
    informer.on('error', () => {
```

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+      replicasetInformer: this.createReplicasetInformer(kc, ns, context),
+    };
+  }
+
+  private createPodInformer(kc: KubeConfig, ns: string, context: KubeContext): Informer<V1Pod> & ObjectCache<V1Pod> {
+    const k8sApi = kc.makeApiClient(CoreV1Api);
+    const listFn = () => k8sApi.listNamespacedPod(ns);
+    const informer = makeInformer(kc, `/api/v1/namespaces/${ns}/pods`, listFn);
+
+    informer.on('add', (_obj: V1Pod) => {
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]:
```suggestion
    informer.on('add', () => {
```

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+      for (const state of this.contextsInternalState.values()) {
+        await state.podInformer?.stop();
+        await state.deploymentInformer?.stop();
+        await state.replicasetInformer?.stop();
+      }
+      this.contextsInternalState.clear();
+      this.contextsState.clear();
+      this.dispatchContextsState();
+    }
+    console.log('==> # informers', this.contextsInternalState.size, this.contextsState.size);
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]: Replace `==>` with meaningful, human readable message.

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+  private restartInformer(informer: Informer<KubernetesObject> & ObjectCache<KubernetesObject>, context: KubeContext) {
+    const kubernetesConfiguration = this.configurationRegistry.getConfiguration('kubernetes');
+    const liveContextsInfo = kubernetesConfiguration.get<boolean>('LiveContextsInfo');
+    if (!liveContextsInfo) {
+      return;
+    }
+
+    informer
+      .start()
+      .then(() => {})
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]: Why is that required?

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+      .catch((err: unknown) => {
+        console.log('==> catched err', err);
```
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]: There should be readable message that would explain where the error has happened.

Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-context-state.ts]:
```suggestion
        console.log('==> catched err', String(err));
```

**Review by dgolovin — CHANGES_REQUESTED** [review]

File: packages/main/src/plugin/kubernetes-context-state.ts (hunk 13 above)
Comment (feloy) [reply packages/main/src/plugin/kubernetes-context-state.ts]: I'm using this `==>` sign to grep into it, to specifically get my logs during debug sessions. Most of these logs will probably disappear before the final revision

File: packages/main/src/plugin/kubernetes-context-state.ts (hunk 14 above)
Comment (feloy) [reply packages/main/src/plugin/kubernetes-context-state.ts]: Not sure it is. Removing it

File: packages/main/src/plugin/kubernetes-context-state.ts (hunk 13 above)
Comment (feloy) [reply packages/main/src/plugin/kubernetes-context-state.ts]: This should be fixed with latest commit

**Comment by feloy** [comment]
Thanks @dgolovin for your review. I've added unit tests, the PR is ready for review again

File: packages/main/src/plugin/kubernetes-client.ts [hunk 16]
```diff
@@ -157,7 +162,7 @@ export class KubernetesClient {
     // add configuration
     const kubeconfigConfigurationNode: IConfigurationNode = {
       id: 'preferences.Kubernetes.Kubeconfig',
-      title: 'Path to the kubeconfig file',
+      title: 'Kubernetes Contexts',
```
Comment (deboer-tim) [inline packages/main/src/plugin/kubernetes-client.ts]: I'm really sorry about the merge conflict I (inadvertently) created. I still think the title should be just 'Kubernetes' here though, at least until we have several other Kubernetes preferences.

**Review by deboer-tim — CHANGES_REQUESTED** [review]
I'm seeing some minor UI problems while testing this:
- 'Replicasets' extends underneath the cluster details on the right.
- When the window width is < 950px, the namespace and its value collide. If you keep shrinking it, cluster has the same problem. There is enough width to show these without overlapping, it just isn't using the right side of the UI.
- My Minikube cluster is old/doesn't exist. However, every ~5s it flashes and I can see 0 pods/deployments/replicasets for a split second before it resets to unreachable.

Screenshot 2024-01-16 at 5 12 44 PM

It also feels a bit odd that we chose Deployments/Ingress/Routes/Services as the first set of objects to add nav pages for, but are showing Deployments/Replicasets here. I wonder if we should use a different set here, or even just not show replicasets for now (which would also avoid the first problem).

**Comment by feloy** [comment]
> * My Minikube cluster is old/doesn't exist. However, every ~5s it flashes and I can see 0 pods/deployments/replicasets for a split second before it resets to unreachable.

The problem is that, when the informer fails starting, we receive a `connect` event with a 
 null error. And we are checking if we a correctly connected to a cluster by checking this same error in the `connect` event. Due to this inconsistency, we have two choices:

- have a small delay (<1s), and change a state from connected<>disconnect after this delay
- wait for a add/update/delete/change event to declare a cluster reachable. But in this case, a cluster with no resource will be declared as unreachable. To fix this we could watch for nodes, for examples, for which we know we will always have a resource, and then always receive at least an `add` event (but the problem is that many users won't have permissions to watch this `node` resource)

**Comment by feloy** [comment]
> > * My Minikube cluster is old/doesn't exist. However, every ~5s it flashes and I can see 0 pods/deployments/replicasets for a split second before it resets to unreachable.
> 
> The problem is that, when the informer fails starting, we receive a `connect` event with a null error. And we are checking if we a correctly connected to a cluster by checking this same error in the `connect` event. Due to this inconsistency, we have two choices:
> 
> * have a small delay (<1s), and change a state from connected<>disconnect after this delay
> * wait for a add/update/delete/change event to declare a cluster reachable. But in this case, a cluster with no resource will be declared as unreachable. To fix this we could watch for nodes, for examples, for which we know we will always have a resource, and then always receive at least an `add` event (but the problem is that many users won't have permissions to watch this `node` resource)

I've added a new commit with the first solution (waiting a short delay before sending the value of the reachable state, and do not send it if a new one it sent later), to avoid flashing

**Comment by deboer-tim** [comment]
I looked through this again - the work is really good and UI just has a couple minor issues, but we've let it stagnate. We already have other PRs like #5850 that are solving some of the same problems. Assuming @feloy is still tied up in other work, I'd like to find someone else who can take over this PR soon and wrap up the (hopefully) last few things.

**Comment by feloy** [comment]
> I looked through this again - the work is really good and UI just has a couple minor issues, but we've let it stagnate. We already have other PRs like https://github.com/containers/podman-desktop/pull/5850 that are solving some of the same problems. Assuming @feloy is still tied up in other work, I'd like to find someone else who can take over this PR soon and wrap up the (hopefully) last few things.

I should be able to finish this PR, could you sumarize the UI changes that would be needed?

File: packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte [hunk 17]
```diff
@@ -95,32 +96,72 @@ async function handleDeleteContext(contextName: string) {
           {/if}
         </div>
         <div class="grow flex-column divide-gray-900 text-gray-400">
-          <div class="text-xs bg-charcoal-800 p-2 rounded-lg mt-1 grid grid-cols-6">
-            <span class="my-auto font-bold col-span-1 text-right">CLUSTER</span>
-            <span class="my-auto col-span-5 text-left pl-0.5 ml-3" aria-label="context-cluster">{context.cluster}</span>
-          </div>
+          <div class="flex flex-row">
+            {#if $kubernetesContextsState.get(context.name)}
```
Comment (dgolovin) [inline packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte:100]: I have current context pointing to sandbox instance and I can query resources using `kubectl`. That means my context state should be reachable, but for some reason I don't see part of UI below.

**Review by dgolovin — CHANGES_REQUESTED** [review]
The part with counters does not show up for active reachable context pointing to sandbox instance.

File: packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte (hunk 17 above)
Comment (feloy) [reply packages/renderer/src/lib/preferences/PreferencesKubernetesContextsRendering.svelte:100]: Did you enable the feature in the preferences, as the feature is disabled by default?

**Comment by feloy** [comment]
As discussed during daily meeting, I've ushed a new commit removing the replicaset information.

I'll split this PR into several smaller ones.

**Comment by feloy** [comment]
Closing in favor of #5912  and following
